sql/core/src/main/scala/org/apache/spark/sql/functions.scala

```diff
@@ -2841,6 +2841,31 @@ object functions {
   // DateTime functions
   //////////////////////////////////////////////////////////////////////////////////////////////
 
+  /**
+   * Creates a datetime interval
+   *
+   * @param years Number of years
+   * @param months Number of months
+   * @param weeks Number of weeks
+   * @param days Number of days
+   * @param hours Number of hours
+   * @param mins Number of mins
+   * @param secs Number of secs
+   * @return A datetime interval
+   * @group datetime_funcs
+   * @since 3.2.0
+   */
+  def make_interval(
+      years: Column = lit(0),
```

Comment:
Can we avoid default argument? This doesn't work in Java.

Reply:
@HyukjinKwon - thank you for the code review.  The default arguments allow for this nice syntax:

```scala
df.withColumn("plus_2_hours", col("first_datetime") + make_interval(hours = lit(2)))
```

Without the default arguments, we'll need to invoke the method with a lot of arguments:

```scala
val twoHourInterval = make_interval(lit(0), lit(0), lit(0), lit(0), lit(2), lit(0), lit(0))
df.withColumn("plus_2_hours", col("first_datetime") + twoHourInterval)
```

We could overload the method and allow for both syntaxes (using a little trick with the last argument that's optional).

```scala
def make_interval(
    years: Column,
    months: Column,
    weeks: Column,
    days: Column,
    hours: Column,
    mins: Column,
    secs: Column,
    failOnError: Boolean): Column = withExpr {
  MakeInterval(
    years.expr, months.expr, weeks.expr, days.expr, hours.expr, mins.expr, secs.expr, failOnError)
}
```

This would allow Scala users to either manually list 8 arguments (e.g. `make_interval(lit(0), lit(0), lit(0), lit(0), lit(2), lit(0), lit(0), true)` or use the clean default syntax (e.g. `make_interval(hours = lit(2))`).  Would this give the Java users a usable interface?

I'm also fine removing the default values if that's what's necessary to make this method compatible with Java.  @MaxGekk - open to your thoughts too cause I know you were originally in favor [of the default arguments approach](https://issues.apache.org/jira/browse/SPARK-33995?focusedCommentId=17258718&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-17258718).

Reply:
What kind of issue will we face? Do we have to specify all arguments when we call such method from Java, or impossible to call it all?

Reply:
@MaxGekk - I added a test to demonstrate that `make_interval` can be called from the Java API (with the Scala method that has default values) when all 7 arguments are passed to the method.  e.g. this Java code works:

```java
Column twoYears = make_interval(lit(2), lit(0), lit(0), lit(0), lit(0), lit(0), lit(0));
Dataset<Row> df = spark.createDataFrame(rows, schema).withColumn("plus_two_years", col("some_date").plus(twoYears));
```

Someone with Java experience should check my result.  If the current implementation works with the Java API, then I think it's ok to keep it, especially because it allows for much more elegant Scala code.

Reply:
> Can we avoid default argument? This doesn't work in Java.

That's why I recommend focusing on one language only and think of others later. For Java we might need more functions to look Scala-ish. No idea what R requires.

Reply:
@MaxGekk @HyukjinKwon - Here's where we currently stand on the default arguments issue:

* Default arguments are not used by other functions in `org.apache.spark.sql.functions`, so exposing a function with default param values isn't consistent with what's been done in the past.
* The default arguments to `make_interval` make the function better for Scala users because it allows for this nice syntax `make_interval(hours = lit(2))`
* Java users can use the `make_interval` function that's defined with the default arguments, [as illustrated in this test](https://github.com/apache/spark/pull/31073/commits/c9492a99ffa8a3ab1380ebde21c1ea200d02c187) as long as they supply all seven arguments

There are good arguments to use and to avoid default arguments with `make_interval`.  Can you please let me know your updated thoughts on if you think we should use the default values or not?  Thank you.

Reply:
> That's why I recommend focusing on one language only and think of others later. For Java we might need more functions to look Scala-ish. No idea what R requires.

While I don't see a huge issue with adding Python and R API separately for this specific function, Scala API often sets limits for what is possible on the other side.  Moreover, we seldom extend JVM API just for the sake of guest languages these days. So unless API is trivial (like here) or uses well established interface (like ML model), it's better to design things at the same time.

Reply:
@zero323 - that's a good point.  I plan on creating additional PRs for some of the other functions that are in the SQL API, but aren't in the Scala/Python APIs yet and I'll just add functions for all the APIs in a single PR going forward.

Do you have any additional feedback for this PR or think it's good to get merged?

Reply:
LGTM @MrPowers